keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
+higher level. This could be added later.
+
+### Phase 2
+
+#### Namespace scoped policy configuration
+
+For phase 1, policy configuration resources were only allowed to be cluster
+scoped. We can support namespace scoped policy configuration as follows:
+
+- If the configuration resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. 
+
+#### CEL expression scoping
+
+CRD validation rules are scoped to the schema at the location in the OpenAPIv3
+where they are defined. This make validation rules far easier to author by
+eliminating the need to dereference from the root of an object to the field that
+needs to be validated. Should we provide something similar?
+
+Use cases:
+
+- Dereferencing to deeply nested fields in CEL. A `scope` field containing a
+  field path be sufficient. E.g. `spec.containers[*].image`
+- More ambitiously, some way to scope a CEL expression to a type found nested in
+  multiple kinds, e.g. "io.k8s.api.core.v1.PodTemplateSpec" or
+  "io.k8s.api.core.v1.Container" would help policy authors apply policies more
+  broadly and uniformly (e.g. to match initContainers, containers,
+  ephemeralContainers).
+
+#### Reporting violations
+
+We need:
+
+- Formatted messages (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?). Note that CRD Validation Rules do NOT support formatted
+  messages (but if we add support for formatted message for admission control,
+  we should consider adding the feature to CRD validation as well).
+- severity (e.g. "Warning")
+- Audit Annotations - Information recorded to the audit system but not reported
+  back in the response.
+- status types (e.g. Forbidden can be returned from a webhook using
+  HTTP 403) (can a match rule override this? or can it be selected via a CEL
+  expression?)
+- Similar to how each match has configuration, can each match also select
+  the severity of messages? It make it possible to configure a policy as
+  warning only when first enabling it. Some policy frameworks support this.
+
+#### kube-apiserver authorizer checks
+
+kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
+as a way of doing things like:
+
+- Validate that only a user with a specific permission can set an enum to the
+  "HOLD" value.
+- Validate that only a controller responsible for a finalizer can remove it from
+  the finalizers field.
+ 
+ Concerns: "Is joe authorized to do this"? That only works for the objects joe
+ creates, but not objects that get created on joe's behalf by a controller.
+ Ditto for updates. I heard someone cite PSP as an example for why it's needed,
```

Comment:
Also used by
1. PodSecurityPolicy (kube)
2. CertificateApproval (kube)
3. CertificateSigning (kube)
4. OwnerReferencesPermissionEnforcement (kube)
5. network.openshift.io/ExternalIPRanger
6. route.openshift.io/IngressAdmission
7. scheduling.openshift.io/PodNodeConstraints
8. network.openshift.io/RestrictedEndpointsAdmission
9. security.openshift.io/SecurityContextConstraint
10. security.openshift.io/SCCExecRestrictions

and that's just the in-tree in openshift, doesn't include other project like kubevirt and open-cluster-management. The pattern is commonly and successfully used both in the kube project (outside of PSP) and by admission consumers elsewhere.

Problems in PSP were multi-faceted and secondary authoriation checks will be required to replace a significant portion of admission plugins.

Reply:
Even if it's not appropriate for all scenarios, I think admission that reframes some attribute of the incoming object into an authorization check would be pretty straightforward to implement in-process and as a webhook, would be pretty simple to express in a cel expression, and would be pretty powerful.

I think it was uniquely problematic for PSP for two reasons:
* PSP guarded pods, which are almost completely created by controllers on behalf of other users. That's not the case for most objects, so there's plenty of object types where authorization checks of an attribute of the object against the requesting user could make sense.
* PSP tried to compensate for the controller-created scenario by _also_ checking the pod's service account authorization, which was _really_ confusing.

Reply:
What are the advantages to using authz over a custom configuration object for these examples?

Another disadvantage to using authz is it means the admission execution is no longer deterministic, and we can't take advantage of the performance analysis for validation.

Reply:
Per offline discussion, we're going to accumulate more use cases and start sketching out a proposal. I'll make some general notes in the KEP in the short term.